lambda-runtime/src/AWS/Lambda/Client.hs

```diff
@@ -0,0 +1,192 @@
+module AWS.Lambda.Client
+  ( HTTPConfig
+  , InternalLambdaClientError (..)
+  , LambdaClient
+  , LambdaResponse (..)
+  , getHttpConfig
+  , getNextLambdaResponse
+  , sendBootError
+  , sendEventError
+  , sendEventSuccess
+  )
+where
+
+import AWS.Prelude
+import Control.Exception (displayException)
+import Control.Monad.Except (ExceptT(..), liftEither, throwError, withExceptT)
+import System.Environment (lookupEnv)
+
+import qualified Data.Aeson          as JSON
+import qualified Data.ByteString     as BS
+import qualified Network.HTTP.Client as HTTP
+import qualified Network.HTTP.Types  as HTTP
+
+data HTTPConfig = HTTPConfig
+  { request :: HTTP.Request
+  , manager :: HTTP.Manager
+  }
+
+data LambdaResponse = LambdaResponse
+  { event     :: JSON.Value
+  , requestId :: RequestId
+  }
+  deriving stock Show
+
+newtype RequestId = RequestId Text
+  deriving stock (Show)
+
+instance Conversion BS.ByteString RequestId where
+  convert (RequestId requestId) = encodeUtf8 requestId
+
+data InternalLambdaClientError
+  = LambdaContextDecodeError Text
+  | MissingLambdaRunTimeApi
+  | InvalidLambdaRunTimeApi Text
+  | ConnectionError HTTP.HttpException
+  | ResponseBodyDecodeFailure Text
+  | UnExpectedRunTimeError HTTP.Status Text
+  | PayLoadTooLarge Text
+  deriving stock Show
+
+instance Exception InternalLambdaClientError
+
+type LambdaClient = ExceptT InternalLambdaClientError IO
+
+getHttpConfig :: LambdaClient HTTPConfig
+getHttpConfig = do
+  awsLambdaRuntimeApi <- ExceptT $ maybeToEither MissingLambdaRunTimeApi
+    <$> lookupEnv "AWS_LAMBDA_RUNTIME_API"
+
+  req <- withExceptT (InvalidLambdaRunTimeApi . showc @Text)
+    . ExceptT
+    . try @_ @HTTP.HttpException
+    $ HTTP.parseRequest ("http://" <> awsLambdaRuntimeApi)
+
+  man <- liftIO $ HTTP.newManager
+    $ HTTP.managerSetProxy HTTP.noProxy
+    $ HTTP.defaultManagerSettings
+        { HTTP.managerResponseTimeout = HTTP.responseTimeoutNone
+        , HTTP.managerConnCount = 1
+        , HTTP.managerIdleConnectionCount = 1
+        }
+  return $ HTTPConfig req man
```

Comment:
__question:__ I'm curious about the use of `return` here. Do you prefer `return` or `pure`? I realize they are the same thing now.

Reply:
The rule of thumb is to always use the minimal powerful tool/abstraction when there are two available options. `return` is more powerful than `pure`, therefore pure would have been a better pick here. Going to update this code.

Reply:
> The rule of thumb is to always use the minimal powerful tool/abstraction when there are two available options.

We have the same rule on the ruby side and used it for a long time. Good it know it exists here too.

FWIW It's called [The Rule of Least Power](https://en.wikipedia.org/wiki/Rule_of_least_power).

Reply:
Nice to know their is formal principle. Thanks for sharing.